haystack/retriever/dense.py

```diff
@@ -552,54 +522,210 @@ def retrieve(self, query: str, filters: dict = None, top_k: Optional[int] = None
+            raise ValueError(f"Unknown retriever embedding model format {model_format}")
+
+
+class _DefaultEmbeddingEncoder(_EmbeddingEncoder):
+
+    def __init__(
+            self,
+            retriever: EmbeddingRetriever
+    ):
+
+        self.embedding_model = Inferencer.load(
+            retriever.embedding_model, revision=retriever.model_version, task_type="embeddings",
+            extraction_strategy=retriever.pooling_strategy,
+            extraction_layer=retriever.emb_extraction_layer, gpu=retriever.use_gpu,
+            batch_size=4, max_seq_len=512, num_processes=0
+        )
+        # Check that document_store has the right similarity function
+        similarity = retriever.document_store.similarity
+        # If we are using a sentence transformer model
+        if "sentence" in retriever.embedding_model.lower() and similarity != "cosine":
+            logger.warning(f"You seem to be using a Sentence Transformer with the {similarity} function. "
+                           f"We recommend using cosine instead. "
+                           f"This can be set when initializing the DocumentStore")
+        elif "dpr" in retriever.embedding_model.lower() and similarity != "dot_product":
+            logger.warning(f"You seem to be using a DPR model with the {similarity} function. "
+                           f"We recommend using dot_product instead. "
+                           f"This can be set when initializing the DocumentStore")
+
+    def embed(self, texts: Union[List[List[str]], List[str], str]) -> List[np.ndarray]:
+        # TODO: FARM's `sample_to_features_text` need to fix following warning -
+        # tokenization_utils.py:460: FutureWarning: `is_pretokenized` is deprecated and will be removed in a future version, use `is_split_into_words` instead.
+        emb = self.embedding_model.inference_from_dicts(dicts=[{"text": t} for t in texts])
+        emb = [(r["vec"]) for r in emb]
+        return emb
+
+    def embed_queries(self, texts: List[str]) -> List[np.ndarray]:
+        return self.embed(texts)
+
+    def embed_passages(self, docs: List[Document]) -> List[np.ndarray]:
+        passages = [d.text for d in docs] # type: ignore
         return self.embed(passages)
+
+
+class _SentenceTransformersEmbeddingEncoder(_EmbeddingEncoder):
+
+    def __init__(
+            self,
+            retriever: EmbeddingRetriever
+    ):
+        try:
+            from sentence_transformers import SentenceTransformer
+        except ImportError:
+            raise ImportError("Can't find package `sentence-transformers` \n"
+                              "You can install it via `pip install sentence-transformers` \n"
+                              "For details see https://github.com/UKPLab/sentence-transformers ")
+        # pretrained embedding models coming from: https://github.com/UKPLab/sentence-transformers#pretrained-models
+        # e.g. 'roberta-base-nli-stsb-mean-tokens'
+        if retriever.use_gpu:
```

Comment:
Can we make it util function it is used at multiple places. Some places we use pytorch conda check as well.

Reply:
Sounds good but that's suitable for another PR, right @lalitpagaria ?

Reply:
Yes agree